.ci/test.sh

```diff
@@ -29,13 +29,20 @@ if [[ "$TASK" == "cpp-tests" ]]; then
     exit 0
 fi
 
-conda create -q -y -n $CONDA_ENV python=$PYTHON_VERSION
+# on Ubuntu 14.04, gpu_source build segfaults with gcc 10.x or 11.x
+INITIAL_CONDA_CONSTRAINTS="python=${PYTHON_VERSION}"
+if [[ $TASK == "gpu" ]] && [[ $OS_NAME == "linux" ]] && [[ $IN_UBUNTU_LATEST_CONTAINER != "true" ]]; then
+    INITIAL_CONDA_CONSTRAINTS="${INITIAL_CONDA_CONSTRAINTS} libgcc-ng=9.3.0 libgomp=9.3.0"
+fi
```

Comment:
I found the builds for `Linux gpu_source` segfault with newer versions of `libgcc-ng` and `libgomp` installed.

With Anaconda default channels, we've been getting v9.3.0 of those dependencies. See, for example, the most recent build on `master` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12029&view=logs&j=9ce42e5d-f31a-544a-a6de-dc42f8c013ed&t=b696def5-78ae-590c-4234-04332342ba2d)).

```text
The following NEW packages will be INSTALLED:

  ...
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.3.0-h5101ec6_17
  libgomp            pkgs/main/linux-64::libgomp-9.3.0-h5101ec6_17
```

Switching to `conda-forge`, `mamba` / `conda` (I tried both) really wanted to install much newer versions of these dependencies. I tried pinning to `<11.0` ([link](https://github.com/microsoft/LightGBM/pull/4953/commits/2455e0d2c06807d0d31435e74d3dc199be2cc951)), and `<10.0` ([link](https://github.com/microsoft/LightGBM/pull/4953/commits/1cac6f8417a190efb41e0e261ff0d3a58d85cf42))...both still resulted in a segmentation fault in the tests in `tests/c_api_test/test_.py`.

I'm not sure what is happening here, but I think we just haven't seen this problem before because Anaconda default channels is giving us such an old version of `libgcc` and `libgomp` for the `Linux gpu_source` job.

Reply:
Ah, I see. It's a pity!
Could you please also check that this isn't elated to the used Python version?

Reply:
On that recent `master` build I linked, I see we got:

```text
pkgs/main/linux-64::python-3.9.7-h12debd9_1
```

On one of the builds where `Linux gpu_source` segfaulted, I see

```text
python             conda-forge/linux-64::python-3.9.9-h543edf9_0_cpython
python_abi         conda-forge/linux-64::python_abi-3.9-2_cp39
```

I'm not sure if the presence of `python_abi` is significant or not. I can try researching this a bit more in an hour or so.

Reply:
Interesting, on the most recent build of this PR (which passed), I see `python_abi` was installed but we got `python` 3.9.6.

([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12031&view=logs&j=9ce42e5d-f31a-544a-a6de-dc42f8c013ed&t=b696def5-78ae-590c-4234-04332342ba2d))

```text
  + python                3.9.6  h49503c6_1_cpython  conda-forge/linux-64      27 MB
  + python_abi              3.9  2_cp39              conda-forge/linux-64       4 KB
```

I'll try a build that pins `python` to 3.9.7 but doesn't pin `libgcc` and `libgomp`.

Reply:
Interesting! Pinning to `python=3.9.6` and not pinning anything else for `Linux gpu_source`, the job succeeded!

([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12040&view=logs&j=9ce42e5d-f31a-544a-a6de-dc42f8c013ed&t=b696def5-78ae-590c-4234-04332342ba2d))

I then tried just removing any special conditions in https://github.com/microsoft/LightGBM/pull/4953/commits/30b5e3972a235b464771f0050943ce46d7ef6176, and the `Linux gpu_source` build failed with a segmentation fault ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12041&view=logs&j=9ce42e5d-f31a-544a-a6de-dc42f8c013ed&t=b696def5-78ae-590c-4234-04332342ba2d)).

I compared the logs of the two jobs using PyCharm's file differ. The table below has the full set of differences in dependencies managed by `mamba`.

| package         | successful build           | failed build               |
|:---------------:|:--------------------------:|:--------------------------:|
|bzip2            | --                         | 1.0.8 (h7f98852_4)         |
|dbus             | 1.13.6 (h48d8840_2)        | 1.13.6 (h5008d03_3)        |
|gettext          | 0.19.8.1 (h0b5b191_1005)   | 0.19.8.1 (h73d1719_1008)   |
|glib             | 2.68.4 (h9c3ff4c_0)        | --                         |
|glib-tools       | 2.68.4 (h9c3ff4c_0)        | --                         |
|graphviz         | 2.49.1 (h85b4f2f_0)        | 2.50.0 (h85b4f2f_1)        |
|gst-plugins-base | 1.18.5 (hf529b03_0)        | 1.18.5 (hf529b03_3)        |
|gstreamer        | 1.18.5 (h76c114f_0)        | 1.18.5 (h9f60fe5_3)        |
|harfbuzz         | 3.0.0 (h83ec7ef_1)         | 3.1.1 (h83ec7ef_0)         |
|libglib          | 2.68.4 (h3e27bee_0)        | 2.70.2 (h174f98d_1)        |
|libffi           | 3.3 (h58526e2_2)           | 3.4.2 (h7f98852_5)         |
|libnsl           | --                         | 2.0.0 (h7f98852_0)         |
|librsvg          | 2.52.2 (hc3c00ef_0)        | 2.52.5 (hc3c00ef_1)        |
|python           | 3.9.6 (h49503c6_1_cpython) | 3.9.9 (h543edf9_0_cpython) |

I also noticed that `libstdcxx-ng` ends up being the same version in both builds, but that it was available as of `mamba create` for the successful one and wasn't included until the later `mamba install` for the failed build.

So I tried just explicitly adding `libstdcxx-ng` as a dependency in `conda create`, in https://github.com/microsoft/LightGBM/pull/4953/commits/c9ba0b285e4b3d33db75acbe0837866791e8a8b4. Unfortunately, the `Linux gpu_source` build still segfaults with that change ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12043&view=logs&j=9ce42e5d-f31a-544a-a6de-dc42f8c013ed&t=b696def5-78ae-590c-4234-04332342ba2d)).

So I think for now, to get this working and unblock CI for the rest of the project, we should just pin that job to python=3.9.6.

Reply:
Oh, sorry for being not clear enough! I meant check different **minor** Python versions, 3.7 and 3.8.

Reply:
Ah ok. I just pushed a commit testing `Linux gpu_source` with several different versions of Python: https://github.com/microsoft/LightGBM/pull/4953/commits/f6f63cbb9b4a9cf138f3580ae4223a8acdd0e94a

See the results at https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12051&view=results. This job succeeds with `PYTHON_VERSION` set to 3.7, 3.8, and 3.9.6. It segfaults with `PYTHON_VERSION` set to 3.9 and 3.10.

![image](https://user-images.githubusercontent.com/7608904/150277614-fbece60c-46b1-4a1e-923e-8378b0d0f952.png)